flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,279 @@
+
+	private static final Time timeout = Time.seconds(10L);
+
+	@Rule
+	public final TemporaryFolder tmp = new TemporaryFolder();
+
+	private final JobID jobId = new JobID();
+	private TestingHighAvailabilityServices haServices;
+	private SettableLeaderRetrievalService jobManagerLeaderRetriever;
+	private TestingRpcService rpc;
+
+	@Before
+	public void setup() {
+		haServices = new TestingHighAvailabilityServices();
+		jobManagerLeaderRetriever = new SettableLeaderRetrievalService();
+		haServices.setResourceManagerLeaderRetriever(new SettableLeaderRetrievalService());
+		haServices.setJobMasterLeaderRetriever(jobId, jobManagerLeaderRetriever);
+		rpc = new TestingRpcService();
+	}
+
+	@Test
+	public void testPartitionReleaseAfterDisconnect() throws Exception {
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> taskExecutorGateway.disconnectJobManager(jobId, new Exception("test")),
+			(jobId, jobAwareShuffleEnvironment) -> assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId)));
+	}
+
+	@Test
+	public void testPartitionReleaseAfterReleaseCall() throws Exception {
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> taskExecutorGateway.releasePartitions(jobId, Collections.singletonList(partitionId)),
+			(jobId, jobAwareShuffleEnvironment) -> assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId)));
+	}
+
+	@Test
+	public void testPartitionReleaseAfterShutdown() throws Exception {
+		// don't do any explicit release action, so that the partition must be cleaned up on shutdown
+		testPartitionRelease(
+			(jobId, partitionId, taskExecutorGateway) -> { },
+			(jobId, jobAwareShuffleEnvironment) -> { });
+	}
+
+	private void testPartitionRelease(
+		TriConsumer<JobID, ResultPartitionID, TaskExecutorGateway> releaseAction,
+		BiConsumer<JobID, JobAwareShuffleEnvironmentImpl> postReleaseActionAssertion) throws Exception {
+
+		final TaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
+
+		final TaskSlotTable taskSlotTable = new TaskSlotTable(
+			Collections.singletonList(ResourceProfile.UNKNOWN),
+			new TimerService<>(TestingUtils.defaultExecutor(), timeout.toMilliseconds()));
+
+		final JobLeaderService jobLeaderService = new JobLeaderService(taskManagerLocation, RetryingRegistrationConfiguration.defaultConfiguration());
+
+		final TaskExecutorLocalStateStoresManager localStateStoresManager = new TaskExecutorLocalStateStoresManager(
+			false,
+			new File[]{tmp.newFolder()},
+			Executors.directExecutor());
+
+		final JobAwareShuffleEnvironmentImpl<?> jobAwareShuffleEnvironment = new JobAwareShuffleEnvironmentImpl<>(new NettyShuffleEnvironmentBuilder().build());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setTaskManagerLocation(taskManagerLocation)
+			.setTaskSlotTable(taskSlotTable)
+			.setJobLeaderService(jobLeaderService)
+			.setTaskStateManager(localStateStoresManager)
```

Comment:
it might be better to adjust `TaskManagerServices` to not default mock `taskSlotTable` and `localStateStoresManager` for benefiting all the tests if you like.

Reply:
yes we should definitely look into that; since this would potentially affect many other tests as well let's move this into a follow-up.